plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -627,92 +663,96 @@ type ServiceAffinity struct {
 	labels        []string
 }
 
-func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) algorithm.FitPredicate {
+func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) (algorithm.FitPredicate, func(pm *predicateMetadata)) {
 	affinity := &ServiceAffinity{
 		podLister:     podLister,
 		serviceLister: serviceLister,
 		nodeInfo:      nodeInfo,
 		labels:        labels,
 	}
-	return affinity.CheckServiceAffinity
+
+	return affinity.CheckServiceAffinity, func(pm *predicateMetadata) {
+		var err error
+		// We can't know at the start what labels we are selecting against.
+		pm.servicePods[""], err = podLister.List(createSelectorFromLabels(nil))
```

Comment:
- Ah yes, since PredicateMetadata is done once per scheduling event i was wondering why it didnt retain the entire pod itself....  I assumed maybe that it was decoupled for a reason, after alll, its using other pod information.  If we preserve the whole pod that definetly simplifeis things.

- Regarding podListers... what do you mean `filed of PredicatreMetadataFactory` ?